### PR TITLE
feat(stepfunctions): allow intrinsic functions for json path

### PR DESCRIPTION
**Pull request by msambol**
`QueryString` can start with an [intrinsic function](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html).

Here is an example of a working `QueryString`:
```
"QueryString.$": "States.Format('select contact_id from interactions.conversation where case when \\'{}\\' is not null then bu=\\'{}\\' else 1=1 end and case when \\'{}\\' is not null then channel=\\'{}\\' else 1=1 end and case when {} is not null then year={} else 1=1 end and case when {} is not null then month={} else 1=1 end and case when {} is not null then day={} else 1=1 end limit 1;', States.JsonToString($.completed), $.partitions.bu, $.partitions.bu, $.partitions.channel, $.partitions.channel, $.partitions.year, $.partitions.year, $.partitions.month, $.partitions.month, $.partitions.day, $.partitions.day)"
```

### CODE REVIEW

**Comment by moofish32** [comment]
@BenChaimberg - is this something you can look at? Without this feature it seems hard to have any dynamic behavior with Query Tasks?

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md [hunk 1]
```diff
@@ -205,6 +205,29 @@ const submitJob = new tasks.LambdaInvoke(this, 'Invoke Handler', {
 });
 ```
 
+You can also use [intrinsic functions](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html) with `JsonPath.stringAt()`. 
+Here is an exmaple of starting an Athena query:
+
+```ts
+const startQueryExecutionJob = new tasks.AthenaStartQueryExecution(this, 'Athena Start Query', {
+  queryString: sfn.JsonPath.stringAt("States.Format('select contacts where year={};', $.year)"),
+),
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:214]: typo
```suggestion
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md [hunk 2]
```diff
@@ -205,6 +205,29 @@ const submitJob = new tasks.LambdaInvoke(this, 'Invoke Handler', {
 });
 ```
 
+You can also use [intrinsic functions](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html) with `JsonPath.stringAt()`. 
+Here is an exmaple of starting an Athena query:
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:209]:
```suggestion
Here is an exmaple of starting an Athena query that is dynamically created using the task input:
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md [hunk 3]
```diff
@@ -205,6 +205,29 @@ const submitJob = new tasks.LambdaInvoke(this, 'Invoke Handler', {
 });
 ```
 
+You can also use [intrinsic functions](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html) with `JsonPath.stringAt()`.
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:208]: I think it's definitely a good thing that we're unblocking the use of intrinsic functions, I'm just hesitant to advertise `JsonPath.stringAt` as the entry point. However, I'm not sure if there's a better way without a full design of how to model intrinsic functions properly within the CDK. Any thoughts on a better way to present this to the user as a stopgap until that design takes place?

File: packages/@aws-cdk/aws-stepfunctions/test/fields.test.ts [hunk 4]
```diff
@@ -71,6 +71,10 @@ describe('Fields', () => {
   }),
   test('datafield path must be correct', () => {
     expect(JsonPath.stringAt('$')).toBeDefined();
+    expect(JsonPath.stringAt('States.Format')).toBeDefined();
+    expect(JsonPath.stringAt('States.StringToJson')).toBeDefined();
+    expect(JsonPath.stringAt('States.JsonToString')).toBeDefined();
+    expect(JsonPath.stringAt('States.Array')).toBeDefined();
 
     expect(() => JsonPath.stringAt('$hello')).toThrowError(jsonPathValidationErrorMsg);
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions/test/fields.test.ts:79]: maybe add some fake intrinsic function here ('States.FooBar')

File: packages/@aws-cdk/aws-stepfunctions/lib/fields.ts [hunk 5]
```diff
@@ -219,8 +219,12 @@ function validateJsonPath(path: string) {
     && path !== '$$'
     && !path.startsWith('$$.')
     && !path.startsWith('$[')
+    && !path.startsWith('States.Format')
+    && !path.startsWith('States.StringToJson')
+    && !path.startsWith('States.JsonToString')
+    && !path.startsWith('States.Array')
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions/lib/fields.ts:225]: I think an array-based condition might be better long term, something like
```suggestion
    && ['Format', 'StringToJson', 'JsonToString', 'Array'].every(fn => !path.startsWith(`States.${fn}`))
```

**Review by BenChaimberg — DISMISSED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md (hunk 3 above)
Comment (msambol) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:208]: @BenChaimberg right, this doesn't feel like the correct long-term solution but I think it's important to unblock this. Can we add a `todo` somewhere in the code or docs?

**Review by BenChaimberg — APPROVED** [review]

**Comment by Joe-Zer0** [comment]
Is there concern that the fix being unintuitive?  Would it be better to create a new function for this?
something like
`stepf.IntrinsicFunction.stringAt("States.JsonToString($)")`
instead of
`stepf.JsonPath.stringAt("States.JsonToString($)")`

**Comment by BenChaimberg** [comment]
Yes, absolutely, we should support this in a more fluent way. However, this change unblocks usage which is more important at this stage

**Comment by singlow** [comment]
Does this change allow me to use `listAt` with intrinsic functions or only `stringAt`?

**Pull request merged**